### ISSUE TITLE
Enhanced UPE - WCPay Compatibility: Saved Payment Methods

### ISSUE THREAD

**Issue by FangedParakeet** [Automattic/woocommerce-payments#5022]
### Description

This is the continuation on the integration of Stripe's new UPE payment flows. Please see [this post](pc2DNy-2e2-p2) for more details. This WCPay compatibility issue is one of many to ensure that existing WCPay functionality is unaffected and remains fully operational after refactoring the UPE to follow the new payment flow. 

This issue is to ensure that saved payment methods remain functioning normally. 

This issue is currently blocked, as we await for Stripe to provide us access to their beta UPE release and while we wait to complete the PoC implementation of this new feature (#5019). 

### Acceptance criteria

* Should be able to complete (shortcode) checkout, while saving payment method for future use.
* Should be able to complete checkout using a previously saved payment method.
* Saved payment method should be stored as tokens to the correct payment gateway.
* Should be able to saved both card and SEPA payment methods as payment tokens.
* Removing saved payment methods from Stripe should remove them from WCPay as well.
* Should be able to add, update, and remove saved payment methods from My Account page.

### Additional context

[Stripe UPE – Options to solve WooPay compatibility and WC Pay loss of control over the payments UI/UX](pc2DNy-2e2-p2)
#4530

**Comment by timur27** [comment]
All the tests were performed with the enabled split UPE with deferred intent creation. 

> ❗ This issue doesn't check saving the SEPA payment method since the plugin currently falls back to the legacy UPE when SEPA is activated. Saving SEPA should be reviewed as part of https://github.com/Automattic/woocommerce-payments/issues/6205


| Use-case     | Result |
| ---      | ---       |
| The CC method can be saved | ✅ |
| Guest user shouldn't be able to save a payment method | ✅         |
| Saved CC appears on the shortcode & Blocks checkout | ✅  |
| Pay with saved card via shortcode & Blocks checkout | ✅ | 
| Saved CCs are listed in _Payment methods_ | ✅ | 
| Can mark saved CC as default | ✅  |
| Can delete saved CC | ✅  |
| Terms are shown to the user when checking the save checkbox | ✅  |
| CC can be saved on the _Pay for order page_ | ✅ | 
| Saved CC can be used to pay for order | ✅ |
| Add CC through _Payment methods_ page | ✅  |
| Only reusable payment methods available on the _My account -> Payment methods_ page | 🟥  |
| Save 3DS card on shortcode & Blocks checkout | ✅ |
| Use saved 3DS card on shortcode & Blocks checkout | ✅ |
| Add 3DS card in _My account_ page | ✅ |
| Use a saved 3DS card to pay for an order/renew a subscription | ✅ |
| CC is saved on the subscription payment | ✅ |
| Renew a subscription with a saved CC | ✅ | 
| Save 3DS as a result of subscription payment | ✅ |
| Change the subscription's payment method to another saved card & renew with it | ✅ |
| With WooPay enabled, the checkbox is changed to a WooPay-related save-user checkbox | ✅ |
| Stripe Link payment method is saved during the subscription payment on shortcode & Blocks | ✅ |
| Stripe Link saved payment method can be used to pay for an order | ✅ |
| Saved Stripe Link method can be set as default/removed in _My account_ page | ✅ |


### Fixing failed use-cases 
1️⃣ Only reusable payment methods available on the _My account -> Payment methods_ page: https://github.com/Automattic/woocommerce-payments/pull/6915

**Comment by frosso** [comment]
Discussed during team meeting & discussed here: p1692022749577519-slack-CU6SYV31A
@FangedParakeet and I were unable to reproduce the original issue.
The changes are still valid, but let's wait on @timur27 to come back after AFK to discuss whether it's needed.

**Comment by FangedParakeet** [comment]
Retested all of the above after merging after merging #6965 and being unable to replicate the issue in #6915. My findings with the deferred intent feature flag enabled are as follows...

| Use-case     | Result |
| ---      | ---       |
| The CC method can be saved | ✅ |
| Guest user shouldn't be able to save a payment method | ✅         |
| Saved CC appears on the shortcode & Blocks checkout | ✅  |
| Pay with saved card via shortcode & Blocks checkout | ✅ | 
| Saved CCs are listed in _Payment methods_ | ✅ | 
| Can mark saved CC as default | ✅  |
| Can delete saved CC | ✅  |
| Terms are shown to the user when checking the save checkbox | ✅  |
| CC can be saved on the _Pay for order page_ | ✅ | 
| Saved CC can be used to pay for order | ✅ |
| Add CC through _Payment methods_ page | ✅  |
| Only reusable payment methods available on the _My account -> Payment methods_ page | ✅  |
| Save 3DS card on shortcode & Blocks checkout | ✅ |
| Use saved 3DS card on shortcode & Blocks checkout | ✅ |
| Add 3DS card in _My account_ page | ✅ |
| Use a saved 3DS card to pay for an order/renew a subscription | ✅ |
| CC is saved on the subscription payment | ✅ |
| Renew a subscription with a saved CC | ✅ | 
| Save 3DS as a result of subscription payment | ✅ |
| Change the subscription's payment method to another saved card & renew with it | ✅ |
| With WooPay enabled, the checkbox is changed to a WooPay-related save-user checkbox | ✅ |
| Stripe Link payment method is saved during the subscription payment on shortcode & Blocks | ✅ |
| Stripe Link saved payment method can be used to pay for an order | ✅ |
| Saved Stripe Link method can be set as default/removed in _My account_ page | ✅ |

All these things are clear, however, I did notice one issue. Stripe Link saved payment methods are labelled incorrectly and incomprehensibly on the blocks checkout. Instead of including the email address related to the token, they are simply labelled "Saved token for woocommerce_payments".

![Incorrect Link label on blocks checkout](https://cdn-std.droplr.net/files/acc_1104206/HhNB3M)
_Incorrectly labelled Link token on blocks checkout_

This issue is explained in more detail and resolved in #7016. Once that solution is integrated, I will consider this issue closed. 🎉
